boot/boot.go

```diff
@@ -273,3 +276,92 @@ func MarkBootSuccessful() error {
 
 	return bl.SetBootVars(m)
 }
+
+// MakeBootable sets up the image filesystem with the given rootdir
+// such that it can be booted. bootWith is a map from the paths in the
+// image seed for kernel and boot base to their snap info. This
+// entails:
+//  - installing the bootloader configuration from the gadget
+//  - creating symlinks for boot snaps from seed to the runtime blob dir
+//  - setting boot env vars pointing to the revisions of the boot snaps to use
+//  - extracting kernel assets as needed by the bootloader
+func MakeBootable(model *asserts.Model, rootdir string, bootWith map[string]*snap.Info, unpackedGadgetDir string) error {
+	if len(bootWith) != 2 {
+		return fmt.Errorf("internal error: MakeBootable can only be called with exactly one kernel and exactly one core/base boot info: %v", bootWith)
+	}
+
+	// install the bootloader configuration from the gadget
+	if err := bootloader.InstallBootConfig(unpackedGadgetDir, rootdir); err != nil {
+		return err
+	}
+
+	// setup symlinks for kernel and boot base from the blob directory
+	// to the seed snaps
+
+	snapBlobDir := dirs.SnapBlobDirUnder(rootdir)
+	if err := os.MkdirAll(snapBlobDir, 0755); err != nil {
+		return err
+	}
+
+	for fn := range bootWith {
+		dst := filepath.Join(snapBlobDir, filepath.Base(fn))
+		// construct a relative symlink from the blob dir
+		// to the seed snap file
+		relSymlink, err := filepath.Rel(snapBlobDir, fn)
+		if err != nil {
+			return fmt.Errorf("cannot build symlink for boot snap: %v", err)
+		}
+		if err := os.Symlink(relSymlink, dst); err != nil {
+			return err
+		}
+	}
+
+	// Set bootvars for kernel/core snaps so the system boots and
+	// does the first-time initialization. There is also no
+	// mounted kernel/core/base snap, but just the blobs.
+	bl, err := bootloader.Find(rootdir, &bootloader.Options{
+		PrepareImageTime: true,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot set kernel/core boot variables: %s", err)
+	}
+
+	m := map[string]string{
+		"snap_mode":       "",
+		"snap_try_core":   "",
+		"snap_try_kernel": "",
+	}
+	if model.DisplayName() != "" {
+		m["snap_menuentry"] = model.DisplayName()
+	}
+
+	for fn, info := range bootWith {
+		bootvar := ""
+
+		switch info.GetType() {
```

Comment:
Should this error our on TypeApp ?

Reply:
it is also not checking that there is exactly one kernel and one base mmh

I'll land and iterate